src/CMakeLists.txt

```diff
@@ -294,15 +289,39 @@ ELSE()
 	MESSAGE(FATAL_ERROR "Asked for unknown SHA1 backend ${SHA1_BACKEND}")
 ENDIF()
 
-# Include PCRE and its POSIX regex compatibility layer when it is required
-IF (HAVE_REGCOMP_L)
+# Specify regular expression implementation
+IF(REGEX STREQUAL "")
+	CHECK_SYMBOL_EXISTS(regcomp_l "regex.h;xlocale.h" HAVE_REGCOMP_L)
+	CHECK_SYMBOL_EXISTS(pcre_regcomp "pcreposix.h" HAVE_PCRE)
+
+	IF (HAVE_REGCOMP_L)
+		SET(REGEX "regcomp_l")
+	ELSEIF (HAVE_PCRE)
+		SET(REGEX "pcre")
+	ELSE ()
+		SET(REGEX "builtin")
+	ENDIF ()
+ENDIF()
+
+IF(REGEX STREQUAL "regcomp_l")
 	ADD_FEATURE_INFO(regex ON "using system regcomp_l")
-ELSE()
+	SET(GIT_REGEX_REGCOMP_L 1)
+ELSEIF(REGEX STREQUAL "pcre")
+	ADD_FEATURE_INFO(regex ON "using system PCRE")
+	SET(GIT_REGEX_PCRE 1)
+ELSEIF(REGEX STREQUAL "regcomp")
+	ADD_FEATURE_INFO(regex ON "using system regcomp")
+	SET(GIT_REGEX_REGCOMP 1)
+ELSEIF(REGEX STREQUAL "builtin")
+	ADD_FEATURE_INFO(regex ON "using bundled PCRE")
```

Comment:
This invocation could be repeated if we instead set a variable. A matter of taste though, please do as you like